### PR TITLE
Strings

### PR DESCRIPTION
**Pull request by RichardShashalevich**

### CODE REVIEW

File: exercises/02_check_palindrom/02_check_palindrom.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+func InSet(b byte, set string) bool {
+	for _, v := range set {
+		if v == rune(b) {
+			return true
+		}
+	}
+	return false
+}
+
+func TrimRight(str string) string {
```
Comment (zayac) [inline exercises/02_check_palindrom/02_check_palindrom.go:22]: Why do you need this function here?

File: exercises/01_trim_right/01_trim_right_test.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+import "testing"
+
+func TestTrimRight(t *testing.T) {
+	for _, tc := range []struct {
+		name  string
+		input string
+		want  string
+	}{
+		{"empty", "", ""},
```
Comment (zayac) [inline exercises/01_trim_right/01_trim_right_test.go:11]: What about a test case "abc "? What about tests with unicode characters?

File: exercises/02_check_palindrom/02_check_palindrom_test.go [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+import "testing"
+
+func TestIsPalidndrom(t *testing.T) {
+	for _, tc := range []struct {
+		name  string
+		input string
+		want  bool
+	}{
+		{"empty", "", true},
```
Comment (zayac) [inline exercises/02_check_palindrom/02_check_palindrom_test.go:11]: Could you add more trivial tests, like "a", "ab"? What about strings with unicode characters?

File: exercises/02_check_palindrom/02_check_palindrom.go [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+	return str
+}
+
+func IsPalidrom(str string) bool {
+	var tmp string
+	for i := len(str) - 1; i >= 0; i-- {
+		tmp += string(str[i])
+	}
+	return tmp == str
+}
```
Comment (zayac) [inline exercises/02_check_palindrom/02_check_palindrom.go:35]: optional: consider implementing a solution without allocating extra memory.

File: exercises/03_count_byte/03_count_byte_test.go [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+func TestCountByte(t *testing.T) {
+	for _, tc := range []struct {
+		name  string
+		str   string
+		sbyte byte
+		want  int
+	}{
+		{"hello", "hello", 0x6c, 2},
+		{"empty", "", 0x6c, 0},
+		{"Latvian letters", "abčdē", 0xc4, 2},
```
Comment (zayac) [inline exercises/03_count_byte/03_count_byte_test.go:14]: In most cases, instead of explicit hex values, please use runes converted to `byte`, e.g. `byte('l')`.

Obviously, you need one or several tests where `sbyte` is a byte that doesn't have a corresponding printable character.

File: exercises/04_count_rune/04_count_rune.go [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+func CountRune(str string, sbyte rune) int {
+	var count int
+	str1 := []rune(str)
+	for i := 0; i < len(str1); i++ {
+		if str1[i] == sbyte {
+			count++
+		}
+	}
+	return count
```
Comment (zayac) [inline exercises/04_count_rune/04_count_rune.go:17]: optional: consider an implementation without allocating extra memory

File: exercises/06_count_words/06_count_words_test.go [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+func TestCountWords(t *testing.T) {
+	for _, tc := range []struct {
+		name string
+		str  string
+		want int
+	}{
+		{"empty", "", 0},
+		{"Hello, world", "hello, world", 2},
+		{"Thats a lot of text", `Java is a language that basically outsourced memory management entirely to its garbage collector. This turned out to be a big mistake. However, to be able to explain why, I need to cover more of the details.
```
Comment (zayac) [inline exercises/06_count_words/06_count_words_test.go:13]: Why do you need a test with a long of text? Why smaller test cases wouldn't work?

File: exercises/06_count_words/06_count_words_test.go [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+import "testing"
+
+func TestCountWords(t *testing.T) {
+	for _, tc := range []struct {
+		name string
+		str  string
+		want int
+	}{
+		{"empty", "", 0},
```
Comment (zayac) [inline exercises/06_count_words/06_count_words_test.go:11]: During the lesson we wrote much more tests for this program. Consider adding them as well.

**Comment by zayac** [comment]
Great commit messages!

File: exercises/01_trim_right/01_trim_right.go [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+func TrimRight(str string) string {
+	for i := len(str) - 1; i > 0 && InSet(str[i-1], trimSet); i-- {
+		str = str[:i-1]
+	}
+	return str
+}
+
+func main() {
+
+	fmt.Print("Enter a string: ")
```
Comment (yarcat) [inline exercises/01_trim_right/01_trim_right.go:31]: nit: It looks a bit strange in the command-line:
```cli
Enter a string: Enter text: 
```
I suggest either adding a string to explain what this program does, or just to drop this line.

File: exercises/01_trim_right/01_trim_right.go [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+	for _, v := range set {
+		if v == rune(b) {
+			return true
+		}
+	}
+	return false
+}
+
+func TrimRight(str string) string {
+	for i := len(str) - 1; i > 0 && InSet(str[i-1], trimSet); i-- {
```
Comment (yarcat) [inline exercises/01_trim_right/01_trim_right.go:23]: This wouldn't work for `str = " "` (any single separator)

File: exercises/04_count_rune/04_count_rune.go [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+)
+
+func CountRune(str string, sbyte rune) int {
+	var count int
+	str1 := []rune(str)
```
Comment (yarcat) [inline exercises/04_count_rune/04_count_rune.go:11]: nit: `for-range` defaults to `rune` for `string` variables:
```go
for _, c := range str {
  if c == sbyte {
    count++
  }
}
```

File: exercises/04_count_rune/04_count_rune.go [hunk 12]
```diff
@@ -0,0 +1,28 @@
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+)
+
+func CountRune(str string, sbyte rune) int {
```
Comment (yarcat) [inline exercises/04_count_rune/04_count_rune.go:9]: nit: I recommend `r` instead of `sbyte`. Why `sbyte`? What does it mean? `r` at least stands for `rune`.
